ec2_eni: Move to boto3, add support for tags and name

healem opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR contains 2 major changes to ec2_eni:

1.  Move the module to boto3 from boto (boto is no longer used by this module)
2.  Add support for tags and add a name field (that uses tag:Name in AWS)

This also adds a new way to find an interface - by name and subnet_id. This will make it
easier to remain idempotent when using dynamic ip addresses.

I've also squashed a minor bug where secondary ip addresses were not included in the
output. #54630 addresses this bug, but in the old boto-dependent code.

This closes #54630 

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ec2_eni

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

Sanity checks run locally against python 2.7, 3.5, and 3.7 using the following command and varying the python version:
`ansible-test sanity -v --docker --python 2.7 ec2_eni`

Manually ran the integration test:
`ansible-test integration ec2_eni --docker --python 3.7`

Manual testing against AWS performed with ansible version:
```
ansible 2.8.0a1.post0
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 3.7.3rc1 (default, Mar 13 2019, 11:01:15) [GCC 8.3.0]
```

<!--- Paste verbatim command output below, e.g. before and after your change -->
Before:
```
 "interface": {
        "description": "myComplexNic descriptn",
        "groups": {
            "sg-0609923bf1372031d": "onlySshAccess"
        },
        "id": "eni-0539ff06598413abc",
        "mac_address": "0a:72:3d:f1:1c:b0",
        "owner_id": "686104308526",
        "private_ip_address": "10.0.84.2",
        "private_ip_addresses": [
            {
                "primary_address": true,
                "private_ip_address": "10.0.84.2"
            }
        ],
        "source_dest_check": true,
        "status": "pending",
        "subnet_id": "subnet-02dc6a04a5460ea5f",
        "vpc_id": "vpc-00fd8af7d3a908c13"
    },
```
After:
```
"interface": {
        "description": "myComplexNic descriptn",
        "groups": {
            "sg-0609923bf1372031d": "onlySshAccess"
        },
        "id": "eni-0e6eb8bf710e26822",
        "mac_address": "0a:24:b8:de:13:e8",
        "name": "myENI-1",
        "owner_id": "686104308526",
        "private_ip_address": "10.0.84.2",
        "private_ip_addresses": [
            {
                "primary_address": true,
                "private_ip_address": "10.0.84.2"
            },
            {
                "primary_address": false,
                "private_ip_address": "10.0.18.102"
            }
        ],
        "source_dest_check": true,
        "status": "available",
        "subnet_id": "subnet-02dc6a04a5460ea5f",
        "tags": {
            "Name": "myENI-1",
            "another": "tag",
            "group": "Finance"
        },
        "vpc_id": "vpc-00fd8af7d3a908c13"
    }
```

healem committed:
Move ec2_eni to boto3, add support for tags and name, add integration test

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -323,46 +397,86 @@ def create_eni(connection, vpc_id, module):
+                connection.delete_network_interface(eni["NetworkInterfaceId"])
                 raise
 
+        eni = uniquely_find_eni(connection, module, eni)
```
If all the other code paths call this, this line is unnecessary.

healem replied:
Good catch.  I'll remove it.

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -378,161 +492,237 @@ def modify_eni(connection, vpc_id, module, eni):
+                connection.assign_private_ip_addresses(
+                    NetworkInterfaceId=eni["NetworkInterfaceId"],
+                    PrivateIpAddresses=secondary_addresses_to_add,
+                    AllowReassignment=allow_reassignment
```
The Amazon docs state that reassignment is asynchronous and you need to wait for it.  Do we want to handle that here?  https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_AssignPrivateIpAddresses.html

healem replied:
Yes, I'll add a wait.

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -542,17 +732,48 @@ def get_sec_group_list(groups):
+
+
+@AWSRetry.jittered_backoff()
+def manage_tags(eni, new_tags, purge_tags, connection):
```
This doesn't exist in module_utils?

healem replied:
No.  I can move it there.  I found this method in ec2_instance, so I can move it to module_utils and refactor that occurance, too.

healem replied:
The tag handling in this function is currently ENI specific.  It could be made more generic, but AWS is inconsistent on where it sets the tags.  For instances its in "Tags" and for ENI it is in "TagSet".  So I punted for now.

jeking3 commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -574,35 +795,28 @@ def main():
+            ['secondary_private_ip_addresses', 'secondary_private_ip_address_count']
+        ],
+        required_if=([
+            ('attached', True, ['instance_id']),
```
The state: absent qualifier was dropped.  Is that ok?

healem replied:
Yes.  Otherwise you can only delete an eni by eni_id.  I dropped it to allows delete of the eni by (name and subnet_id), (subnet_id and private_ip_address), or (instance_id and device_index).

healem replied:
Leaving this conversation open - its worth getting feedback from others on this point.

jeking3 reviewed (commented):
A few questions... also this needs an integration test (I know it didn't have one before though).

jeking3 commented:
You could take the integration test from #54630 as a start for integration testing.

healem commented:
I added the integration test from https://github.com/ansible/ansible/pull/54630 and extended it.  I also addressed the feedback provided by @jeking3

healem commented:
The failure is the same failure seen in https://github.com/ansible/ansible/pull/54630 regarding permissions in the CI system.

```
"msg": "Failed to create eni shippable-118040-102-eni for subnet-08ef34b3891cda01d in vpc-03fbae6810df95fc6 with None: An error occurred (UnauthorizedOperation) when calling the CreateNetworkInterface operation: You are not authorized to perform this operation."
```

jeking3 commented:
I can confirm these changes work and provide idempotency if you specify the name and subnet_id.

jeking3 reviewed (changes requested):
When I try to delete an eni by id that isn't there, it says: {"msg": "Failed to delete ENI: could not find ENI with params {'eni_id': 'eni-0e8c90e60720481ab'...}

Fixing...

jeking3 committed:
Fix idempotent delete in ec2_eni

jeking3 reviewed (approved):
Resolved the delete issue I found.

jeking3 commented:
shipit

healem committed:
Update version_added to 2.9

caphrim007 committed:
Added missing kwargs

s-hertel commented on test/integration/targets/ec2_eni/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/aws
+shippable/aws/group2
```
To help get these tests running in CI, you can find the permissions necessary to run those tests, such as with https://docs.ansible.com/ansible/devel/dev_guide/platforms/aws_guidelines.html#aws-permissions-for-integration-tests, and then open a pull request for https://github.com/mattclay/aws-terminator with the things that aren't already covered.

Please also add these to the hacking/aws_config/testing_policies/ directory.

caphrim007 replied:
@s-hertel the policy in hacking is quite different from the policies in aws-terminator. Terminator has a LOT more. Does terminator need to be reflected in the hacking policies?

s-hertel commented:
@wimnat Do you mind reviewing this?

s-hertel commented:
The integration tests are now passing, looks like it's just a sanity failure because version_added should be bumped to 2.10.

caphrim007 commented:
I'll update this in a little bit when I get in front of a computer

caphrim007 committed:
Update parameter version to 2.10

caphrim007 committed:
Change version to string

caphrim007 commented:
@s-hertel ready_for_review

flowerysong commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -574,35 +838,28 @@ def main():
+    )
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 required for this module')
```
Missing boto3 is automatically handled by AnsibleAWSModule.

caphrim007 replied:
removed

caphrim007 committed:
Remove redundant check for boto3

caphrim007 commented:
@healem if you have a moment to review what's here (my changes included) I'd appreciate it so-as to merge

caphrim007 commented:
@wimnat would you please review this?

mike-healey commented on lib/ansible/modules/cloud/amazon/ec2_eni.py:
```diff
@@ -378,161 +534,238 @@ def modify_eni(connection, vpc_id, module, eni):
+    if eni_id is None and private_ip_address is None and (instance_id is None and device_index is None):
+        return None
+
+    filter_set = False
```
filter_set seems redundant - all of the "if" statements below can just check filters directly.

caphrim007 replied:
@mike-healey it looks like that boolean is used to determine if the function should return `None` or not. However, it looks like one could equally accomplish that by just checking for the size of the `filters` list because that boolean can only be set when(at least) one filter is set.

Would you concur?

caphrim007 replied:
@mike-healey I've changed the code to address your comments

mike-healey reviewed (approved):
Just a minor suggestion.  Otherwise LGTM.

caphrim007 committed:
Remove redundant usage of variables

s-hertel committed:
Add a more thorough test suite for ec2_eni

s-hertel committed:
Add a waiter for the ec2_eni module

Fix defaults for tags and purge_tags

Fix exception handling

Simplify some logic by using get()

Fix changed when modifying tags

Fix attaching instances

s-hertel committed:
Sanity fixes

s-hertel committed:
add a changelog

s-hertel reviewed (approved):
I added a more thorough test suite and fixed a number of bugs it unearthed that worked on the boto version. It's worth noting that I've changed the defaults for `tags` and `purge_tags` to be more similar to most of the AWS modules (`purge_tags` [guidelines](https://docs.ansible.com/ansible/devel/dev_guide/platforms/aws_guidelines.html#dealing-with-tags)).

Since I added tests to exercise more code paths this will need another CI update, but besides that, looks safe to merge to me. @healem, @caphrim007, @jeking3 - If any of you have time to review the changes, I'd appreciate it.

tremble commented on test/integration/targets/ec2_eni/tasks/main.yaml:
```diff
@@ -0,0 +1,499 @@
+  - name: create a subnet
+    ec2_vpc_subnet:
+      cidr: "{{ vpc_cidr }}"
+      az: "{{ aws_region }}a"
```
There's no guarantee AZa will be available: If you spin up a new account and try to use us-west-1 you get 2 of the 3 randomly.

better would be to use aws_az_info

s-hertel replied:
Good point 👍

tremble commented on test/integration/targets/ec2_eni/tasks/main.yaml:
```diff
@@ -0,0 +1,499 @@
+  - assert:
+      that:
+        - result.changed
+        - result.interface.private_ip_addresses | length == 1
```
We know the IP address we expect, why not test that too?

tremble commented on test/integration/targets/ec2_eni/tasks/main.yaml:
```diff
@@ -0,0 +1,499 @@
+      that:
+        - result.changed
+        - result.interface.id == eni_id_1
+        - result.interface.private_ip_addresses | length == 3
```
Should we test that we actually have IP addresses here?
```
  - result.interface.private_ip_addresses[0] | ipaddr
  - result.interface.private_ip_addresses[1] | ipaddr
  - result.interface.private_ip_addresses[2] | ipaddr

mkrizek commented:
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/amazon.aws.

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project
